frontend/pkg/frontend/node_pool.go

```diff
@@ -0,0 +1,404 @@
+		nodePoolDoc.SystemData = systemData
+		docUpdated = true
+	}
+
+	// Here the difference between a nil map and an empty map is significant.
+	// If the Tags map is nil, that means it was omitted from the request body,
+	// so we leave any existing tags alone. If the Tags map is non-nil, even if
+	// empty, that means it was specified in the request body and should fully
+	// replace any existing tags.
+	if hcpNodePool.TrackedResource.Tags != nil {
+		nodePoolDoc.Tags = hcpNodePool.TrackedResource.Tags
+		docUpdated = true
+	}
+
+	if docUpdated {
+		err = f.dbClient.SetResourceDoc(ctx, nodePoolDoc)
+		if err != nil {
+			f.logger.Error(fmt.Sprintf("failed to upsert document for %s: %v", nodePoolResourceID, err))
+			arm.WriteInternalServerError(writer)
+			return
+		}
+		f.logger.Info(fmt.Sprintf("document upserted for %s", nodePoolResourceID))
+	}
+
+	err = f.StartOperation(writer, request, operationRequest, nodePoolDoc.InternalID)
+	if err != nil {
+		f.logger.Error(fmt.Sprintf("failed to write operation document: %v", err))
+		arm.WriteInternalServerError(writer)
+		return
+	}
+
+	responseBody, err := marshalCSNodePool(csNodePool, nodePoolDoc, versionedInterface)
+	if err != nil {
+		f.logger.Error(err.Error())
+		arm.WriteInternalServerError(writer)
+		return
+	}
+
+	writer.WriteHeader(successStatusCode)
+
+	_, err = writer.Write(responseBody)
+	if err != nil {
+		f.logger.Error(err.Error())
+	}
+}
+
+func (f *Frontend) DeleteNodePool(writer http.ResponseWriter, request *http.Request) {
+	ctx := request.Context()
+
+	versionedInterface, err := VersionFromContext(ctx)
+	if err != nil {
+		f.logger.Error(err.Error())
+		arm.WriteInternalServerError(writer)
+		return
+	}
+
+	resourceID, err := ResourceIDFromContext(ctx)
+	if err != nil {
+		f.logger.Error(err.Error())
+		arm.WriteInternalServerError(writer)
+		return
+	}
+
+	f.logger.Info(fmt.Sprintf("%s: DeleteNodePool", versionedInterface))
+
+	doc, err := f.dbClient.GetResourceDoc(ctx, resourceID)
+	if err != nil {
+		if errors.Is(err, database.ErrNotFound) {
+			f.logger.Error(fmt.Sprintf("node pool document cannot be deleted -- document not found for %s", resourceID))
+			writer.WriteHeader(http.StatusNoContent)
+		} else {
+			f.logger.Error(fmt.Sprintf("failed to fetch document for %s: %v", resourceID, err))
+			arm.WriteInternalServerError(writer)
+		}
+		return
+	}
+
+	err = f.clusterServiceConfig.DeleteCSNodePool(ctx, doc.InternalID)
+	if err != nil {
+		f.logger.Error(fmt.Sprintf("failed to delete node pool %s: %v", resourceID, err))
+		arm.WriteInternalServerError(writer)
+		return
+	}
+
+	err = f.StartOperation(writer, request, database.OperationRequestDelete, doc.InternalID)
+	if err != nil {
+		f.logger.Error(fmt.Sprintf("failed to write operation document: %v", err))
+		arm.WriteInternalServerError(writer)
+		return
+	}
+
+	writer.WriteHeader(http.StatusAccepted)
+}
+
+func (f *Frontend) OperationResult(writer http.ResponseWriter, request *http.Request) {
```

Comment:
This function shouldn't be in node_pools.go.  It's an asynchronous operation endpoint, not node pool specific.